tests/two_d_array_ut/two_d_array_ut.c

```diff
@@ -0,0 +1,642 @@
+
+    //clean
+    TWO_D_ARRAY_ASSIGN(uint32_t)(&tdarr, NULL);
+}
+
+/* Tests_SRS_TWO_D_ARRAY_07_012: [ TWO_D_ARRAY_FREE_ROW(T) shall free the memory associated with the row specified by row_index and set it to NULL. ]*/
+/* Tests_SRS_TWO_D_ARRAY_07_013: [ TWO_D_ARRAY_FREE_ROW(T) shall return zero on success. ]*/
+TEST_FUNCTION(TWO_D_ARRAY_FREE_ROW_of_THANDLE_succeeds)
+{
+    //arrange
+    int result;
+    TWO_D_ARRAY(THANDLE(A_TEST)) tdarr = TWO_D_ARRAY_CREATE(THANDLE(A_TEST))(5, 5);
+    int add_row_result = TWO_D_ARRAY_ALLOCATE_NEW_ROW(THANDLE(A_TEST))(tdarr, 0);
+    ASSERT_IS_NOT_NULL(tdarr->row_arrays[0]);
+    umock_c_reset_all_calls();
+    STRICT_EXPECTED_CALL(free(IGNORED_ARG));
+
+    //act
+    result = TWO_D_ARRAY_FREE_ROW(THANDLE(A_TEST))(tdarr, 0);
+
+    //assert
+    ASSERT_IS_NOT_NULL(tdarr);
+    ASSERT_ARE_EQUAL(int, add_row_result, 0);
+    ASSERT_ARE_EQUAL(int, result, 0);
+    ASSERT_IS_NULL(tdarr->row_arrays[0]);
+
+    ASSERT_ARE_EQUAL(char_ptr, umock_c_get_expected_calls(), umock_c_get_actual_calls());
+
+    //clean
+    TWO_D_ARRAY_ASSIGN(THANDLE(A_TEST))(&tdarr, NULL);
+}
+
+/* TWO_D_ARRAY_GET(T) */
+
+/* Tests_SRS_TWO_D_ARRAY_07_019: [ If two_d_array is NULL, TWO_D_ARRAY_GET(T) shall fail return NULL. ]*/
+TEST_FUNCTION(TWO_D_ARRAY_GET_fails_when_handle_is_null)
+{
+    //arrange
+    uint32_t* row;
+
+    //act
+    row = TWO_D_ARRAY_GET(uint32_t)(NULL, 0);
+
+    //assert
+    ASSERT_IS_NULL(row);
+    ASSERT_ARE_EQUAL(char_ptr, umock_c_get_expected_calls(), umock_c_get_actual_calls());
+
+    //clean
+}
+
+/* Tests_SRS_TWO_D_ARRAY_07_020: [ If row_index is equal or greater than row_size, TWO_D_ARRAY_GET(T) shall fail return NULL. ]*/
+TEST_FUNCTION(TWO_D_ARRAY_GET_fails_when_row_index_exceeds)
+{
+    //arrange
+    uint32_t* row_result;
+    TWO_D_ARRAY(uint32_t) tdarr = TWO_D_ARRAY_CREATE(uint32_t)(5, 5);
+    int add_row_result = TWO_D_ARRAY_ALLOCATE_NEW_ROW(uint32_t)(tdarr, 0);
+    ASSERT_IS_NOT_NULL(tdarr);
+    ASSERT_ARE_EQUAL(int, add_row_result, 0);
+    ASSERT_IS_NOT_NULL(tdarr->row_arrays[0]);
+    umock_c_reset_all_calls();
+
+    //act
+    row_result = TWO_D_ARRAY_GET(uint32_t)(tdarr, 5);
+
+    //assert
+    ASSERT_IS_NULL(row_result);
+    ASSERT_ARE_EQUAL(char_ptr, umock_c_get_expected_calls(), umock_c_get_actual_calls());
+
+    //clean
+    TWO_D_ARRAY_ASSIGN(uint32_t)(&tdarr, NULL);
+}
+
+/* Tests_SRS_TWO_D_ARRAY_07_021: [ If the array stored in row_index is NULL, TWO_D_ARRAY_GET(T) shall fail and return NULL. ]*/
+TEST_FUNCTION(TWO_D_ARRAY_GET_fails_when_row_not_initialized_yet)
+{
+    //arrange
+    uint32_t* row_result;
+    TWO_D_ARRAY(uint32_t) tdarr = TWO_D_ARRAY_CREATE(uint32_t)(5, 5);
+    int add_row_result = TWO_D_ARRAY_ALLOCATE_NEW_ROW(uint32_t)(tdarr, 0);
+    ASSERT_IS_NOT_NULL(tdarr);
+    ASSERT_ARE_EQUAL(int, add_row_result, 0);
+    ASSERT_IS_NOT_NULL(tdarr->row_arrays[0]);
+    umock_c_reset_all_calls();
+
+    //act
+    row_result = TWO_D_ARRAY_GET(uint32_t)(tdarr, 1);
```

Comment:
>GET [](http://example.com/codeflow?start=29&length=3)

This is really get_row. We should name it like that

Reply:
changed to get_row here and everywhere else